protocol/contracts/v2/access/TempleElevatedAccess.sol

```diff
@@ -31,55 +31,89 @@ abstract contract TempleElevatedAccess is ITempleElevatedAccess {
+        rescuer = initialRescuer;
+        executor = initialExecutor;
     }
 
     /**
      * @notice Set the contract into or out of rescue mode.
      * Only the rescuers are allowed to set.
      */
     function setRescueMode(bool value) external override {
-        if (!rescuers[msg.sender]) revert CommonEventsAndErrors.InvalidAccess();
+        if (msg.sender != rescuer) revert CommonEventsAndErrors.InvalidAccess();
         emit RescueModeSet(value);
         inRescueMode = value;
     }
 
     /**
-     * @notice Grant `account` the emergency operations role
+     * @notice Proposes a new Rescuer.
+     * Can only be called by the current rescuer.
      */
-    function setRescuer(address account, bool value) external override onlyElevatedAccess {
-        if (account == address(0)) revert CommonEventsAndErrors.InvalidAddress(account);
-        emit RescuerSet(account, value);
-        rescuers[account] = value;
+    function proposeNewRescuer(address account) external override {
+        if (msg.sender != rescuer) revert CommonEventsAndErrors.InvalidAccess();
+        if (account == address(0)) revert CommonEventsAndErrors.InvalidAddress();
+        emit NewRescuerProposed(rescuer, _proposedNewRescuer, account);
+        _proposedNewRescuer = account;
     }
 
     /**
-     * @notice Grant `account` the executor role
+     * @notice Caller accepts the role as new Rescuer.
+     * Can only be called by the proposed rescuer
      */
-    function setExecutor(address account, bool value) external override onlyElevatedAccess {
-        if (account == address(0)) revert CommonEventsAndErrors.InvalidAddress(account);
-        emit ExecutorSet(account, value);
-        executors[account] = value;
+    function acceptRescuer() external override {
+        if (msg.sender != _proposedNewRescuer) revert CommonEventsAndErrors.InvalidAccess();
+        emit NewRescuerAccepted(rescuer, msg.sender);
+        rescuer = msg.sender;
+        delete _proposedNewRescuer;
     }
 
     /**
-     * @notice Grant `allowedCaller` the rights to call the function determined by the selector `fnSelector`
+     * @notice Proposes a new Executor.
+     * Can only be called by the current executor or rescuer (if in resuce mode)
+     */
+    function proposeNewExecutor(address account) external override onlyElevatedAccess {
+        if (account == address(0)) revert CommonEventsAndErrors.InvalidAddress();
+        emit NewExecutorProposed(executor, _proposedNewExecutor, account);
+        _proposedNewExecutor = account;
+    }
+
+    /**
+     * @notice Caller accepts the role as new Executor.
+     * Can only be called by the proposed executor
+     */
+    function acceptExecutor() external override {
+        if (msg.sender != _proposedNewExecutor) revert CommonEventsAndErrors.InvalidAccess();
+        emit NewExecutorAccepted(executor, msg.sender);
+        executor = msg.sender;
+        delete _proposedNewExecutor;
```

Comment:
I'm not sure that it is more efficient for an address actually - just tested this in Remix

unset() ==> 5233 gas
rm() ==> 5164 gas

```
pragma solidity ^0.8.18;
// SPDX-License-Identifier: AGPL-3.0-or-later

contract DeleteVsSet {
    address private _proposedNewExecutor;

    function set() external {
        _proposedNewExecutor = address(this);
    }

    function unset() external {
        _proposedNewExecutor = address(0);
    }

    function rm() external {
        delete _proposedNewExecutor;
    }
}
```

Reply:
Called set() twice, then unset(), then set() twice again, then rm()

Reply:
Looking at the opcodes, when doing:
```
_proposedNewExecutor = address(0);
```

There's an extra `PUSH1 00`, probably because it needs to put the `address(0)` on the stack

Reply:
I didn't mean effective by gas efficiency. I meant `delete` operation clears the storage (in addition to setting the primitive value to default), and therefore the position of the variable's storage slot may change by the next transaction. Nothing risky about this because we're not using fixed assembly `SLOT` values. Just bringing it to your attention.

Reply:
Hmm delete changes the SLOT number? Could you give a concrete example? This is new to me :-)

Reply:
Docs here: https://docs.soliditylang.org/en/latest/types.html#delete

Doesn't mention anything about changing storage representation for primatives.

Reply:
I read it from this [article](https://medium.com/stackanatomy/an-ultimate-guide-to-variables-in-solidity-958d4d114bff), under functions. but on second look at the docs i don't see anything concrete on reassigning slots.

Reply:
Hmm yeh that isn't clear at all. It would be very strange for the slot to change there.

Perhaps it's a badly written way of saying
```
x=5;
delete x; //=0;
x=3;
```

Costs more in gas? Idk

Reply:
above snippet does look more costly.
I will take some time later to read yellow paper/jellowpaper/other articles if I find something.